File: internal/cli/workspace.go
package cli

import (
	"fmt"
	"io"
	"slices"
	"strings"

	"github.com/krowkcom/cli/internal/api"
	"github.com/krowkcom/cli/internal/config"
	"github.com/krowkcom/cli/internal/output"
	"github.com/krowkcom/cli/internal/runctx"
)

// workspacesList answers both workspace questions at once: which keys this
// machine holds, and which of them a command run right here would use. The
// second half is the one that changes with the directory, which is why it is
// computed rather than read back from anywhere.
func workspacesList(w io.Writer, f flags, format output.Format, env runctx.Env, colour bool) error {
	ws, source, err := resolveWorkspace(f, env)
	if err != nil {
		return err
	}

	stored := api.StoredWorkspaces()
	view := output.Workspaces{Stored: stored, Resolved: ws, Source: source,
		Shadowed: env("KROWK_TOKEN") != ""}
	// Nothing asked by name means the stored default answers, and the listing
	// should say so rather than showing a default mark the reader has to
	// cross-reference against an empty "resolved" line.
	if ws == "" {
		for _, k := range stored {
			if k.Default {
				view.Resolved, view.Source = k.Name, "stored default"
			}
		}
	}
	fmt.Fprintln(w, output.WorkspaceList(view, format, f.quiet, colour))
	return nil
}

// workspacesUse repoints the machine-wide default at another stored key. It is
// selection, not authentication: it can only name a key `auth login` already
// stored, which is what makes it safe to run without confirming anything.
//
// With no argument and a person at the terminal, the store's contents become
// a picker — the answer was always going to be one of those names, so asking
// beats failing and having them run the listing to copy one out. Everywhere
// else, no argument stays the error it always was.
func workspacesUse(w io.Writer, args []string, f flags, format output.Format, env runctx.Env,
	colour, isTTY bool) error {
	if len(args) == 0 {
		name, err := askForWorkspace("Which workspace should be the default?", f, format, env, isTTY,
			api.Fail("no_workspace",
				"pass the workspace: `krowk workspaces use <name>` — `krowk workspaces` lists them"))
		if err != nil {
			return err
		}
		args = []string{name}
	}
	path, err := api.SetDefaultWorkspace(args[0])
	if err != nil {
		// The store's own message already lists what is there or says the store
		// is empty, so it is surfaced as written rather than paraphrased.
		return api.Fail("unknown_workspace", err.Error())
	}
	fmt.Fprintln(w, output.DefaultWorkspace(args[0], path, format, f.quiet, colour))
	return nil
}

// configShow reports the effective configuration and where each value came
// from. A malformed file still fails it, even though a broken config is
// exactly what someone running this is often chasing: Load cannot say what the
// unreadable file would have set, so the honest answer is the error naming the
// file — which is the diagnosis, delivered as a failure.
func configShow(w io.Writer, f flags, format output.Format, env runctx.Env, colour bool) error {
	cfg, err := config.Load("", env, config.Overrides{Workspace: f.workspace})
	if err != nil {
		return api.Fail("bad_config", err.Error())
	}

	view := output.ConfigView{
		Workspace:  cfg.Workspace,
		Sources:    cfg.Sources,
		GlobalPath: config.GlobalPath(),
	}
	if repo, inRepo := config.RepoPath(""); inRepo {
		view.RepoPath = repo
	}
	fmt.Fprintln(w, output.ConfigShow(view, format, f.quiet, colour))
	return nil
}

// configSet writes one key into a config file: the repository's by default,
// because pinning the repo is what the command exists for, or the global file
// behind --global. Outside a git repository there is no repo file to write, and
// guessing a directory would pin something other than what was meant.
//
// `config set workspace` with the value left off asks, when there is a person
// to ask — the value is one of the stored names, and the picker offers exactly
// those. Other keys have no store to offer from, so their missing value stays
// an error everywhere.
func configSet(w io.Writer, args []string, f flags, format output.Format, env runctx.Env,
	colour, isTTY bool) error {
	missing := api.Fail("missing_argument",
		"pass the key and the value: `krowk config set workspace <name>` — "+
			"keys: "+strings.Join(config.KnownKeys(), ", "))
	if len(args) == 0 {
		return missing
	}
	if len(args) == 1 {
		if args[0] != "workspace" {
			return missing
		}
		name, err := askForWorkspace("Which workspace should this pin?", f, format, env, isTTY, missing)
		if err != nil {
			return err
		}
		args = append(args, name)
	}
	key, value := args[0], args[1]
	if err := knownConfigKey(key); err != nil {
		return err
	}

	path, err := configFile(f)
	if err != nil {
		return err
	}
	if err := config.Set(path, key, value); err != nil {
		return api.Fail("config_unwritable", "could not write "+path+": "+err.Error())
	}
	fmt.Fprintln(w, output.ConfigWrote(key, value, path, format, f.quiet, colour))
	return nil
}

// configUnset removes one key from a config file, same file selection as
// configSet. Removing a key that was never set is success — the state asked
// for is the state there is.
func configUnset(w io.Writer, args []string, f flags, format output.Format, colour bool) error {
	if len(args) == 0 {
		return api.Fail("missing_argument",
			"pass the key: `krowk config unset workspace` — keys: "+strings.Join(config.KnownKeys(), ", "))
	}
	if err := knownConfigKey(args[0]); err != nil {
		return err
	}

	path, err := configFile(f)
	if err != nil {
		return err
	}
	if err := config.Unset(path, args[0]); err != nil {
		return api.Fail("config_unwritable", "could not write "+path+": "+err.Error())
	}
	fmt.Fprintln(w, output.ConfigWrote(args[0], "", path, format, f.quiet, colour))
	return nil
}

// configFile is which file `config set` and `config unset` write: the
// repository's, or the global one when --global asks for it.
func configFile(f flags) (string, error) {
	if f.global {
		return config.GlobalPath(), nil
	}
	repo, inRepo := config.RepoPath("")
	if !inRepo {
		return "", api.Fail("not_in_a_repository",
			"the repo config lives at <git-root>/.krowk/config.json and there is no git root here — "+
				"run it inside the repository, or pass --global for the machine-wide file")
	}
	return repo, nil
}

// askForWorkspace turns a missing argument into a question when a person is
// there to answer it; anywhere else it answers with `otherwise`, the error
// the command would have raised anyway. The store's emptiness is checked
// before the terminal is touched: a picker over nothing is a question with no
// answers, and the fix for it is a login, not a selection.
func askForWorkspace(title string, f flags, format output.Format, env runctx.Env,
	isTTY bool, otherwise error) (string, error) {
	if !interactive(f, format, env, isTTY) {
		return "", otherwise
	}
	stored := api.StoredWorkspaces()
	if len(stored) == 0 {
		return "", api.Fail("not_authenticated",
			"no keys are stored to pick from — `krowk auth login` adds one")
	}
	return pickWorkspace(title, stored)
}

// knownConfigKey rejects a key this build has no idea of, before any file is
// touched. Writing it anyway would succeed and then do nothing, which reads as
// krowk ignoring the user — worse than saying no.
func knownConfigKey(key string) error {
	if slices.Contains(config.KnownKeys(), key) {
		return nil
	}
	return api.Fail("unknown_config_key",
		"`"+key+"` is not a configuration key — keys: "+strings.Join(config.KnownKeys(), ", "))
}

File: crates/krowk/src/cli/workspace.rs
//! `config` and `workspaces`: which workspace a command lands in, and the
//! stored keys it could land in.

use super::{interactive, Ctx};
use crate::config;
use crate::output::workspace::{self as render, ConfigView, Workspaces};
use krowk_api::creds::{self, WorkspaceKey};
use krowk_api::{fail, Error};

/// Which workspace this command was pointed at, and by whom. Empty means
/// nothing asked — "use the stored default", which only the credential store
/// can resolve. A config file that cannot be read is a refusal: somebody wrote
/// it meaning to steer uploads.
pub(crate) fn resolve_workspace(ctx: &Ctx) -> Result<(String, String), Error> {
    let cfg = config::load("", ctx.io.env, &ctx.f.workspace)
        .map_err(|e| fail("bad_config", format!("{e} — fix the file or remove it; `krowk config show` names every layer")))?;
    let source = cfg.sources.get("workspace").cloned().unwrap_or_default();
    Ok((cfg.workspace, source))
}

pub(crate) fn workspaces_list(ctx: &mut Ctx) -> Result<(), Error> {
    let (ws, source) = resolve_workspace(ctx)?;
    let stored = creds::stored_workspaces();
    let mut view = Workspaces { resolved: ws.clone(), source, shadowed: !ctx.env("KROWK_TOKEN").is_empty(), ..Workspaces::default() };
    if let Some(k) = stored.iter().find(|k| k.default).filter(|_| ws.is_empty()) {
        view.resolved = k.name.clone();
        view.source = "stored default".into();
    }
    if !view.resolved.is_empty() && !view.shadowed && creds::resolve_token(ctx.io.env, &ws).is_err() {
        view.key_missing = true;
    }
    view.stored = stored;
    let rendered = render::workspace_list(&view, ctx.format, ctx.f.quiet, ctx.colour);
    ctx.emit(&rendered)
}

pub(crate) fn workspaces_use(ctx: &mut Ctx, args: &[String]) -> Result<(), Error> {
    if creds::stored_workspaces().is_empty() {
        return Err(fail("not_authenticated", "no keys are stored to choose between — `krowk auth login` adds one"));
    }
    let name = match args.first() {
        Some(name) => name.clone(),
        None => ask_for_workspace(
            ctx,
            "Which workspace should be the default?",
            fail("no_workspace", "pass the workspace: `krowk workspaces use <name>` — `krowk workspaces` lists them"),
        )?,
    };
    let path = creds::set_default_workspace(&name).map_err(|e| fail("unknown_workspace", e))?;
    let rendered = render::default_workspace(&name, &path, ctx.format, ctx.f.quiet, ctx.colour);
    ctx.emit(&rendered)
}

pub(crate) fn config_show(ctx: &mut Ctx) -> Result<(), Error> {
    let cfg = config::load("", ctx.io.env, &ctx.f.workspace).map_err(|e| fail("bad_config", e))?;
    let view = ConfigView {
        workspace: cfg.workspace,
        sources: cfg.sources,
        global_path: config::global_path().display().to_string(),
        repo_path: config::repo_path("").map(|p| p.display().to_string()).unwrap_or_default(),
    };
    let rendered = render::config_show(&view, ctx.format, ctx.f.quiet, ctx.colour);
    ctx.emit(&rendered)
}

pub(crate) fn config_set(ctx: &mut Ctx, args: &[String]) -> Result<(), Error> {
    let missing = || {
        fail(
            "missing_argument",
            format!(
                "pass the key and the value: `krowk config set workspace <name>` — keys: {}",
                config::known_keys().join(", ")
            ),
        )
    };
    let (key, value) = match args {
        [] => return Err(missing()),
        [key] if key != "workspace" => return Err(missing()),
        [key] => (key.clone(), ask_for_workspace(ctx, "Which workspace should this pin?", missing())?),
        [key, value, ..] => (key.clone(), value.clone()),
    };
    config::known(&key).map_err(|e| fail("unknown_config_key", e))?;
    let path = config_file(ctx)?;
    config::set(&path, &key, &value).map_err(|e| fail("config_unwritable", format!("could not write {}: {e}", path.display())))?;
    let rendered = render::config_wrote(&key, &value, &path.display().to_string(), ctx.format, ctx.f.quiet, ctx.colour);
    ctx.emit(&rendered)
}

pub(crate) fn config_unset(ctx: &mut Ctx, args: &[String]) -> Result<(), Error> {
    let Some(key) = args.first() else {
        return Err(fail(
            "missing_argument",
            format!("pass the key: `krowk config unset workspace` — keys: {}", config::known_keys().join(", ")),
        ));
    };
    config::known(key).map_err(|e| fail("unknown_config_key", e))?;
    let path = config_file(ctx)?;
    config::unset(&path, key).map_err(|e| fail("config_unwritable", format!("could not write {}: {e}", path.display())))?;
    let rendered = render::config_wrote(key, "", &path.display().to_string(), ctx.format, ctx.f.quiet, ctx.colour);
    ctx.emit(&rendered)
}

/// The file a write goes to: the machine's under --global, else the
/// repository's, and a refusal outside one.
fn config_file(ctx: &Ctx) -> Result<std::path::PathBuf, Error> {
    if ctx.f.global {
        return Ok(config::global_path());
    }
    config::repo_path("").ok_or_else(|| {
        fail(
            "not_in_a_repository",
            "the repo config lives at <git-root>/.krowk/config.json and there is no git root here — run it inside the \
             repository, or pass --global for the machine-wide file",
        )
    })
}

/// A person at a terminal picks from the stored keys; anyone else gets
/// `otherwise`, since a question with nobody to answer it is a hang.
fn ask_for_workspace(ctx: &Ctx, title: &str, otherwise: Error) -> Result<String, Error> {
    if !interactive(ctx) {
        return Err(otherwise);
    }
    let stored = creds::stored_workspaces();
    if stored.is_empty() {
        return Err(fail("not_authenticated", "no keys are stored to pick from — `krowk auth login` adds one"));
    }
    pick_workspace(title, &stored)
}

fn pick_workspace(title: &str, stored: &[WorkspaceKey]) -> Result<String, Error> {
    let labels: Vec<String> = stored
        .iter()
        .map(|k| {
            let mut label = if k.workspace_name.is_empty() { k.name.clone() } else { format!("{}  —  {}", k.workspace_name, k.name) };
            if k.default {
                label += "  (default)";
            }
            label
        })
        .collect();
    let picked = inquire::Select::new(title, labels.clone())
        .prompt()
        .map_err(|_| fail("selection_cancelled", "nothing was selected and nothing was changed"))?;
    let at = labels.iter().position(|l| *l == picked).expect("picked from the list");
    Ok(stored[at].name.clone())
}
